test/meta.cpp

```diff
@@ -743,6 +761,15 @@ TEST_F(Meta, MetaDtor) {
     ASSERT_EQ(empty_type::counter, 0);
     ASSERT_TRUE(dtor.invoke(empty));
     ASSERT_EQ(empty_type::counter, 1);
+
+    dtor = meta::resolve<empty_type*>().dtor();
```

Comment:
I'm looking at this from GitHub, not fully tested yet (you know, daily jobs and the other things that steal time to the open source :smile:).

The goal here is to get the meta type for `empty_type` when I invoke `resolve<empty_type *>`?  
Therefore, pointers type are in a sense no more resolvable? I mean, there won't be any difference between `T` and `T*` from the point of view of the meta system when it comes to returning meta types?

Reply:
Not exactly. Their `meta::type`s are different.

`resolve<empty_type *>() != resolve<empty_type>()`

But their "accessors" (`ctor`, `dtor`, `func`, etc...) are the same. So although their `meta::type` are different, the usage of these different types are very streamlined.

`resolve<empty_type *>().func("my_func") == resolve<empty_type>().func("my_func")`

If I want to check if a given type models the same "class" you need to check if their `clazz` is equal. The `clazz` is defined as always returning the bare user-defined type of a type. That is, the type without any "decoration" like const or pointer. This is the same type that you use when calling `meta::factory` methods.

`resolve<empty_type *>().clazz() == resolve<empty_type>().clazz()`

The `func` accessor for a pointer type behaves as-if:

```
meta::func meta::type::func(const char* name) {
  return clazz().func(name);
}
```

TLDR: If you want to find if a `meta::any` is has an invokable `foo`, you don't have to care about it's concrete type, whether it's const, volatile, pointer or value. It will transparently for any type.

```
meta::any v = ...; // can be T or T*, doesn't matter
v.type().func("foo").invoke(v);
meta::resolve<T>().func("foo").invoke(v); // again, it doesn't matter if `v` is really a `T` here as long as it's `clazz()` is `T`
meta::resolve<T*>().func("foo").invoke(v); // same for `T*`

Reply:
A branch would be great. I was going to suggest that yesterday but forgot about it.

I am not completely satisfied with #3 as it is either, maybe with the branch we can work it out better.

Just ping me whenever you got the branch ready.

Reply:
I've merged it. However, [it fails](https://ci.appveyor.com/project/skypjack/meta/builds/24769080) on appveyor. :cry:

Reply:
> 
> 
> I've merged it. However, [it fails](https://ci.appveyor.com/project/skypjack/meta/builds/24769080) on appveyor. 😢

Huh. That's funny. I tested it with MSVC. Is it using 2017 or 2019?

Reply:
Visual Studio 15 2017.

Reply:
> 
> 
> Visual Studio 15 2017. It fail because of the `reinterpret_cast` within `ptr`. I don't get exactly what's its purpose anyway.

`ptr` creates a uniform representation of the type. So that both `T` and `T*` always return a `void*` that can be `reinterpret_cast`ed back to a `T*` (this is what allows you to call a `func` with either type). This will also be helpful when adding references later on, so that we simply take a pointer and cast into a `void*`.

Haven't looked into detail yet, but looks like a broken `if constexpr` from VS 2017. I'm installing the VS2017 toolchain to see what exactly is the problem.

Reply:
Wait a moment. It's conceptually wrong anyway, literally UB in C++. `T` can be a pointer but also a member pointer, you cannot just cast it to a `void *`. Using a `reinterpret_cast` here can silent the compiler but isn't technically correct.

Reply:
> 
> 
> Wait a moment. It's conceptually wrong anyway, literally UB in C++. `T` can be a pointer but also a member pointer, you cannot just cast it to a `void *`. Using a `reinterpret_cast` here can silent the compiler but isn't technically correct.

I haven't considered the possibility of `T` being a member pointer, but I don't think this will invoke UB. Here's why:

Previously the library always assumed the `void*` given by `any` was a `T*` (since it could only deal with values this was fine). If `T` was a pointer, it returned a `T**`. This allowed `func_node` and other nodes to always assume a `T*` was given and it would dereference it.

This is similar, but now it returns a `T*` for all types (T* returns is returned as-is, just casting into a void*). What `ptr` does is:

If `T` is not a pointer, it returns: `reinterpret_cast<void*>(&value)` (cast a `T*` into `void*`, which is valid)
If `T` is a pointer, it returns `reinterpret_cast<void*>(value)`.

Now, upon use (by `func_node` for example), it can safely assume that the value returned by `ptr` is always a `T*` and can act appropriately:

If the argument is a `T*`, it `reinterpret_cast<T*>(void_ptr)` and forwards to the call. This is equivalent to:
```c++
int* n;
void* np = reinterpret_cast<void*>(&n); // this is the `ptr` function
int* n2 = reinterpret_cast<int*>(np); // this is the `func_node` argument
```

If the argument is a `T`, it `*reinterpret_cast<T*>(void_ptr)` and forwards to the call. This is equivalent to:
```c++
int n;
void* np = reinterpret_cast<void*>(&n); // this is the `ptr` function
int n2 = *reinterpret_cast<int*>(np); // this is the `func_node` argument
```

At no point (unless for an unknown bug) there's a `reinterpret_cast` with invalid types (`T` into `T*` or vice-versa). 

I am now not sure if member pointers can be cast into a `void*`. Also, can you store a member pointer into a `any`? This function will only be called when working with type-erased `any` types.

Reply:
The problem isn't the `handle` nor the `any` class. The problem is the definition of `ptr` within `resolve`, the one that performs the `reinterpret_casts`. Granted that anything bad will happen as long as it's not invoked but still it's easy to observe that there pass both pointers to function and pointers to member, two things that cannot be cast to `void *`.  
Ironically, both gcc and clang silently ignore the fact that we are performing an invalid cast but i'ts true that it's a `reinterpret_cast` and in this case the compilers usually adopt an _you certainly know what you're doing_ policy. This doesn't make those cast valid unfortunately.

>I am now not sure if member pointers can be cast into a void*. Also, can you store a member pointer into a any? This function will only be called when working with type-erased any types.

They cannot and yes, you can store a member pointer into an `any`. The size is potentially bigger than that of a `void *` and the `any` reserves enough space to store it in the dynamic storage while skipping the SBO optimization. This is safe anyway, because what you obtain as an instance is a pointer to something (in this case, a pointer to a pointer to member). The first pointer is a straight pointer, it can be cast to `void *`.

---

Btw, the branch for _it's not a pointer_ is equivalent to `return ptr;`. :-)

Reply:
Oh, I get it now. I think we'll need a special case for member pointers then. A pointer to the member pointer should be fine I think.



> 
> 
> The problem isn't the `handle` nor the `any` class. The problem is the definition of `ptr` within `resolve`, the one that performs the `reinterpret_cast`s. Granted that anything bad will happen as long as it's not invoked but still it's easy to observe that there pass both pointers to function and pointers to member, two things that cannot be cast to `void *`.
> Ironically, both gcc and clang silently ignore the fact that we are performing an invalid cast but i'ts true that it's a `reinterpret_cast` and in this case the compilers usually adopt an _you certainly know what you're doing_ policy. This doesn't make those cast valid unfortunately.
> 
> > I am now not sure if member pointers can be cast into a void*. Also, can you store a member pointer into a any? This function will only be called when working with type-erased any types.
> 
> They cannot and yes, you can store a member pointer into an `any`. The size is potentially bigger than that of a `void *` and the `any` reserves enough space to store it in the dynamic storage while skipping the SBO optimization. This is safe anyway, because what you obtain as an instance is a pointer to something (in this case, a pointer to a pointer to member). The first pointer is a straight pointer, it can be cast to `void *`.
> 
Oh, I get it now. Yeah, you are right. I think we should handle it as as special case and use a pointer to a member pointer, which should be the same length as `void*` then.

> Btw, the branch for _it's not a pointer_ is equivalent to `return ptr;`. :-)

Yeah, I believe the any already holds a `T*` in it's `instance` member. So it's actually the opposite of what I said.

A `any` holding a pointer has `any::instance` pointing to a `T**` so we must cast it, and then dereference it to get a `T*`. 

A `any` holding a value has `any::instance` pointing to `T*` already, so we do a simple pass through. This makes it's behavior equivalent to what it did before.

Reply:
>I think we should handle it as as special case and use a pointer to a member pointer, which should be the same length as void* then.

In fact, you must exclude both pointers to members and pointers to functions. The latter cannot be converted to `void *` as well.  
However, I'd flip the problem on its head and check that `T` is a pointer and `std::remove_pointer_t<T>` isn't a function (or that it's an actual class, probably this is the best bet, what you want is to treat pointers to classes transparently after all). Note that `std::is_pointer_v` is already false for members.  
In this case, you can use a `static_cast` rather than a `reinterpret_cast`, so that the compiler is allowed to tell you that you aren't attempting a wrong conversion. In all the other cases, `return ptr;` works just fine and you don't need casts at all.

Reply:
I mean [this](https://github.com/skypjack/meta/blob/b00f3f53fda2834a37a4ec563e5481213a33977a/src/meta/meta.hpp#L2275), even though VS doesn't like it as well. Let me check into the error.

Reply:
Yeah, it's apparently vs2015 that [goes crazy](https://ci.appveyor.com/project/skypjack/meta/builds/24780014) with this guard. :cry:  
I'll make some other attempts in the afternoon, let me know if you find a work around for it in the meantime.

Reply:
VS2017 `if constexpr` was pretty broken from what I remember, it sometimes would try to evaluate the false branch and (rightfully) fails to compile. It was hit and miss though.

Maybe we'll need to either drop VS2017 or use templates classes and specializations to deal with this the old way. Gotta say, I hate working with MSVC, it is getting better over time, but damn, it is still annoying to work with.

I am pretty sure function pointers are be convertible to `void*`. In both Windows and POSIX `dlsym` (and the equivalent for Win32) returns a `void*` that points to a function in the dynamic library. So unless both POSIX and Win32 are both relying on UB here, I am pretty sure that is fine in this case.

The real problem is that member pointers are usually implemented as an offset relative to the object absolute memory address (and maybe the vtable) and thus might or might not be the same size as a pointer. Function pointers through don't have this problem and are guaranteed to be the same length as a pointer and convertible to a `void*`.

I will see if I can find a solution for that in the weekend. I already installed MSVC 19.16 to test this using the same compiler in appveyor.

Reply:
> 
> 
> I mean [this](https://github.com/skypjack/meta/blob/b00f3f53fda2834a37a4ec563e5481213a33977a/src/meta/meta.hpp#L2275), even though VS doesn't like it as well. Let me check into the error.

Wouldn't a `!std::is_member_object_pointer_v ` and `!is_member_function_pointer` be easier and clearer?

Reply:
Hi @Rogiel - have you found the time to look into this?

Reply:
No, not yet. Sorry.